src/connection.rs

```diff
@@ -416,6 +403,33 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Connection<T> {
                     Either::Right(io)
                 };
 
+            match next_io_event.await? {
+                IoEvent::OutboundReady => {
+                    // Only unpause the control command receiver if we're not
+                    // shutting down already.
+                    if let Shutdown::NotStarted = self.shutdown {
+                        self.control_receiver.unpause();
+                    }
+                }
+                IoEvent::Inbound(frame) => {
+                    if let Some(stream) = self.on_frame(frame).await? {
+                        self.flush_nowait().await.or(Err(ConnectionError::Closed))?;
+                        return Ok(Some(stream))
+                    }
```

Comment:
```suggestion
                    }
                    continue // The sink is still pending
```
Testing with small TCP send buffers showed that this version still stalls `tests/concurrent.rs`, the reason being this missing `continue`. Otherwise we may continue below with a pending write.

Reply:
It may make sense to to change `tests/concurrent.rs` to configure a `16KiB` TCP send buffer for both server and client via  https://docs.rs/tokio/1.2.0/tokio/net/struct.TcpSocket.html. That way the test should catch these issues.

Reply:
> Testing with small TCP send buffers showed that this version still stalls tests/concurrent.rs, the reason being this missing continue. Otherwise we may continue below with a pending write.

Done in https://github.com/romanb/yamux/pull/1/commits/912317e827e4953266322285aab606d24507abcf.

> It may make sense to change tests/concurrent.rs to configure a 16KiB TCP send buffer for both server and client via TcpSocket::set_send_buffer_size. That way the test should catch these issues.

Done in https://github.com/romanb/yamux/pull/1/commits/95167a639becb753a3247ec94e240173ef46ef14. Instead of always setting the send buffer size, it uses quickcheck to alternate between setting it to 16KiB and leaving it unchanged. I would hope quickcheck to prove useful in the future when adding additional variables. Let me know in case you think this is a bit of excessive and always setting the buffer size is good enough.

I am unable to reproduce the stall with the 16KiB send buffer size and without the `continue` statement. I might be missing something. Did it occur regularly, or only after a large amount of retries? I am not questioning the necessity for the `continue` statement.

Reply:
> I am unable to reproduce the stall with the 16KiB send buffer size and without the continue statement. I might be missing something. Did it occur regularly, or only after a large amount of retries?

It likely also depends on the receive buffer sizes. I noticed that I used a configuration of a 128KiB receive buffer for the testing. The OS default maximum receive buffer size is probably a few MiBs. Maybe the test should also randomise the receive buffer sizes, e.g. to always be some random multiple of the send buffer size.

Reply:
> Did it occur regularly, or only after a large amount of retries?

With 16KiB send buffer and `128KiB` receive buffer, not on every run, but at least 1 or 2 out of 10.

Reply:
> Maybe the test should also randomise the receive buffer sizes, e.g. to always be some random multiple of the send buffer size.

Sounds good. Done via https://github.com/romanb/yamux/pull/1/commits/007f86da23d6744e75dfece064c5748785949ebd.

> > Did it occur regularly, or only after a large amount of retries?
> 
> With `16KiB` send buffer and `128KiB` receive buffer, not on every run, but at least 1 or 2 out of 10.

:+1: Was able to reproduce the stall with `Some(TcpBuferSizes { send: 16384, recv: 65536 })`.